---
title: "Mouse miRNA Denoising Algorithm Demo"
author: "Hannah Swan"
date: "November 4, 2024"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE)
```

```{r}
#clear environment
rm(list=ls())

```

```{r, include=FALSE}
#load necessary packages:
library(tidyverse)
library(DESeq2)
library(Biostrings)

#load functions:
source("correct_technical_length_variant_functions.R")
source("correct_isomiR_counts_v2_FUNCTION.R")
```

#### Load Ernesto's benchmark dataset of sequencing mouse miRNAs:
```{r}
load("/scratch/mmccall2_lab/miRNA_reference/Rdata/summarized_experiment_miRNA.RData")
miRNA_se_object = se_object
load("/scratch/mmccall2_lab/miRNA_reference/Rdata/summarized_experiment_isomiR.RData")
seq_level_se_object = se_object
rm(se_object)

miRNA_counts = cbind(rowData(miRNA_se_object), count = miRNA_se_object@assays@data$counts[,"1L"]) %>% data.frame()


num_mirnas = dim(miRNA_se_object)[1]
num_samples = dim(miRNA_se_object)[2]

cat("\nBenchmark dataset contains", num_mirnas, "miRNAs from", num_samples, "samples.\n")

```

Extract rowdata from `seq_level_se_object`
```{r}
rowdata = rowData(seq_level_se_object) %>% data.frame()
#renaming miRNA column for compatibility w/ denoising function 
colnames(rowdata)[2] = 'miRNA_name'
head(rowdata)
```

Extract countdata from `seq_level_se_object`
```{r}
countdata = seq_level_se_object@assays@data$counts %>% data.frame()
head(countdata)

```

#### Select a suitable test miRNA for demonstration 
For each miRNA, filter the rowdata so we are looking only at sequences mapping to that miRNA and then count the number of rows in that dataframe. We want a miRNA with a decent level of expression so relatively high read count but not the highest number of isomiRs mapping to that miRNA because then the algorithm will take awhile. 
```{r}
num_isomirs = vector()
for(miRNA in miRNA_counts$miRNA){
  num_isomirs = c(num_isomirs, filter(rowdata, miRNA_name == miRNA) %>% nrow())
}

miRNA_counts$num_isomirs = num_isomirs
head(miRNA_counts)

mirnas_for_testing = filter(miRNA_counts, count >=1000) %>% filter(., num_isomirs < 1000) %>% select(., miRNA) %>% unlist()

cat("\nmiRNAs suitable for testing are:\n")
filter(miRNA_counts, miRNA %in% mirnas_for_testing)


```

Select a test miRNA
```{r}
test_miRNA = mirnas_for_testing[1]
cat("miRNA we will be using for testing is", test_miRNA, "\n")
```

First we run for just 2 iterations to make sure that algorithm works without any errors:
```{r}
tst = correct_isomiR_counts_v2(test_mirna = test_miRNA, rowdata=rowdata, countdata=countdata, sample_idx = 2, max_iterations = 2, OMEGA_A = 0.05)
```

After one iteration, we have the following `partition_df` object:
```{r}
head(tst$partition_df)
```

The two partitions in the data have the following center sequences with associated read counts in sample 2:
```{r}
filter(tst$partition_df, center == 1)
```

Observe that for this dataset, at least in this sample, we have a lot less degradation in this sample. 
```{r}
filter(tst$partition_df, seqlen == min(seqlen))
```

So only 1 sequence has a length of  15 mapping to this miRNA. This may have to do with the parameters that were used to perform the alignment but the point still stands that we don't have any of the short little 8/9  nucleotide fragments that we had in the ERCC dataset. 

### Allow the algorithm to remove all technical length isomiRs for the specified test miRNA
```{r}
tst_rmv_all = correct_isomiR_counts_v2(test_mirna = test_miRNA, rowdata=rowdata, countdata=countdata, sample_idx=2, max_iterations = 20, OMEGA_A=0.05)
```


```{r}
cat("Results of running  algorithm on miRNA", test_miRNA, ":\n")
table(tst_rmv_all$partition)
cat("There are", max(tst_rmv_all$partition), "true sequences mapping to miRNA",  test_miRNA, "in this sample. \n")
cat("\n The 16 true sequences mapping to miRNA", test_miRNA, "are\n")
filter(tst_rmv_all, center == 1 )
```

```{r}
colnames(countdata)=str_remove_all(colnames(countdata), "X")
colnames(countdata)

liver_samples = sapply(1:5, function(x) return(paste0(x, "L", collapse="")))
liver_sample_indices = which(colnames(countdata) %in% liver_samples)
```

Now we're going to remove all technical isomiRs mapping to the test miRNA as identifed by our model across all pure liver samples:
```{r, eval=FALSE}
tst_rmv_all_across_liver_samples = lapply(liver_sample_indices, function(x) correct_isomiR_counts_v2(test_miRNA, rowdata=rowdata, countdata=countdata, x, max_iterations=20, OMEGA_A=0.05))


```


First look at what the initial center sequence is for each of the 5 samples:
```{r}
initial_center_sequences = vector()
for(i in 1:length(tst_rmv_all_across_liver_samples)){
  cat("The initial center sequence in liver sample", i, "is:\n")
  cs =  filter(tst_rmv_all_across_liver_samples[[i]], center == 1 & partition ==  1) %>% select(.,  uniqueSequence) %>% unlist() 
  print(cs)
  initial_center_sequences = c(initial_center_sequences, cs)
  
}
if(length(unique(initial_center_sequences))==1){
  cat("All samples begin with same true sequence.\n")
}
```

Now look at number of partitions in each sample:
```{r}
num_partitions_by_sample = vector()
for(i in 1:length(tst_rmv_all_across_liver_samples)){
  num = max(tst_rmv_all_across_liver_samples[[i]]$partition)
  num_partitions_by_sample = c(num_partitions_by_sample, num)
}

cat("Each sample has the following number of partitions, which equates to the number of true isomiR sequences\n")
print(num_partitions_by_sample)
```

From this we can definitely see that there is some variability in the number of partitions that exists within this test miRNA for each sample. I suspect that this has to do with varying sequencing depths across each of these samples in combination with partitions that are centered by a sequence with low counts i.e. some later partitions are formed by sequences with only a single observed read. In samples with a greater sequencing depth we might see more of these singleton sequences or even sequences with a handful of counts that get their own partition. \n
I think what's most important 